### PR TITLE
Add configuration options

### PR DESCRIPTION
**Pull request by bondjimbond**
This PR is based on yours, @mjordan, so it could be merged into your PR before yours gets merged into master.

I'd consider this a first step - getting the config functionality in - before the second step of using auto-generated CModel lists and automatically-populated DSIDs.

Note that the default for islandora_usage_stats_callbacks_cmodel_dsids is absurdly long; suggestions for how to shorten it on line 105 would be welcomed.

### CODE REVIEW

**Comment by mjordan** [comment]
@bondjimbond testing some other stuff at the moment, can test later today.

@bryjbrown you cool with this new functionality?

**Comment by bryjbrown** [comment]
@mjordan @bondjimbond I'm okay with the new functionality in theory, but in practice I think this may not be the best way to do it. Having a textarea where you enter 'islandora:cmodel|DSID' pairs seems very likely to introduce bugs when users inevitably spell a cmodel or DSID wrong, or have space after the DSID or something similar. The part where this list gets exploded twice is also kinda hard to follow as well.

My suggestion would be to have the admin menu show you a list of cmodel names, and provide a dropdown select list for each one populated by the datastreams you would most likely want to track. Each cmodel ends up getting its own variable (something like `islandora_usage_stats_callbacks_download_dsid_video` or similar), and this variable gets used in the associative array at https://github.com/mjordan/islandora_usage_stats_callbacks/blob/master/islandora_usage_stats_callbacks.module#L105.

**Comment by bondjimbond** [comment]
@bryjbrown that's exactly what I'd like to do - I just don't know how to do it. Goal is to get all the cmodels with a function, then somehow take that list, put that into a form, and allow users to select a DSID.

If you've got some code snippet ideas you could throw in there as a review, it'd be much appreciated.

**Comment by mjordan** [comment]
@bryjbrown I agree that this type of admin option management is prone to typos, but the reason I :+1: to this style is that I've implemented them in a some of my own modules for admin settings that are updated very infrequently, like this one. Also, the risk of typos is likely be low if the impacted functionality is read-only (e.g., in this case, the number of hits is not reported properly but incorrect data is not written to the db or objects because of the admin setting). Maybe we could start off with this type of form and if the underlying functionality proves to be popular, someone could reimplement the form in a more conventional way.

Your call of course, but I thought I'd offer my perspective.

File: islandora_usage_stats_callbacks.module [hunk 1]
```diff
@@ -102,24 +102,15 @@ function islandora_usage_stats_callbacks_get_object_ids($pid) {
...
-                        'islandora:newspaperCModel' => 'PDF');
- 
+  $dsids = variable_get(islandora_usage_stats_callbacks_cmodel_dsids,'ir:citationCModel|PDF\nir:thesisCModel|PDF\nislandora:sp_pdf|OBJ\nislandora:sp_basic_image|OBJ\nislandora:sp_large_image_cmodel|OBJ\nislandora:binaryObjectCModel|OBJnislandora:sp_videoCModel|MP4\nislandora:sp-audioCModel|PROXY_MP3nislandora:collectionCModel|FALSE\nislandora:compoundCModel|FALSE\nislandora:newspaperIssueCModel|PDF\nislandora:newspaperPageCModel|OBJ\nislandora:sp_web_archive|OBJ\nislandora:sp_disk_image|OBJ\nislandora:pageCModel|OBJ\nislandora:bookCModel|PDF\nislandora:newspaperCModel|PDF');
+  $cmodel_dsids = array();
+
+  // Convert configured DSIDs into an array
+  foreach(explode('\n',$dsids) as $pair){
+    $x=explode('|',$pair);
+    $cmodel_dsids[$x[0]]=$x[1];
+  }
```
Comment (bryjbrown) [inline islandora_usage_stats_callbacks.module]: For this code block, please make sure to add spaces between arguments and assignment operators to make it more readable. 

I'd also request a name change to the variable `$x` for readability as well. Something like 
`list($cmodel, $dsid) = explode('|',$pair);`
is a more readable option.

File: islandora_usage_stats_callbacks.module [hunk 2]
```diff
@@ -102,24 +102,15 @@ function islandora_usage_stats_callbacks_get_object_ids($pid) {
...
-                        'islandora:compoundCModel' => FALSE,
-                        'islandora:newspaperIssueCModel' => 'PDF',
-                        'islandora:newspaperPageCModel' => 'OBJ',
-                        'islandora:sp_web_archive' => 'OBJ',
-                        'islandora:sp_disk_image' => 'OBJ',
-                        'islandora:pageCModel' => 'OBJ',
-                        'islandora:bookCModel' => 'PDF',
-                        'islandora:newspaperCModel' => 'PDF');
- 
+  $dsids = variable_get(islandora_usage_stats_callbacks_cmodel_dsids,'ir:citationCModel|PDF\nir:thesisCModel|PDF\nislandora:sp_pdf|OBJ\nislandora:sp_basic_image|OBJ\nislandora:sp_large_image_cmodel|OBJ\nislandora:binaryObjectCModel|OBJnislandora:sp_videoCModel|MP4\nislandora:sp-audioCModel|PROXY_MP3nislandora:collectionCModel|FALSE\nislandora:compoundCModel|FALSE\nislandora:newspaperIssueCModel|PDF\nislandora:newspaperPageCModel|OBJ\nislandora:sp_web_archive|OBJ\nislandora:sp_disk_image|OBJ\nislandora:pageCModel|OBJ\nislandora:bookCModel|PDF\nislandora:newspaperCModel|PDF');
```
Comment (bryjbrown) [inline islandora_usage_stats_callbacks.module]: I'd strongly recommend reusing the variable from the `$default_dsids` variable from the form (or more properly, store the variable in a utility file or something and call the variable from both places), instead of copypasting the string in here. If you decide to change the default cmodel array, you'll have to change it two places. If you just reference the variable, you'll only have to change it in one place.

**Comment by mjordan** [comment]
Along those lines, if there are plans to move this to Labs, it would be good if everyone could run local code style checks (e.g., `drush dcs`) on all commits.

File: islandora_usage_stats_callbacks.module (hunk 2 above)
Comment (bondjimbond) [reply islandora_usage_stats_callbacks.module]: Definitely! I just don't know how. :)  I mainly learned to code by stealing bits I've seen in other places. So - how to make that variable (which exists in the admin form - $default_dsids) available in this file?

Comment (bryjbrown) [reply islandora_usage_stats_callbacks.module]: You could create a function that returns this associative array of default DSIDs in the .module file, and then use Drupal's [module_load_include](https://api.drupal.org/api/drupal/includes%21module.inc/function/module_load_include/7.x) function in the admin form file. This would load up all the functions from the .module file into memory, allowing you to call that function again and get the default DSIDs.

@mjordan Does that sound right to you, or do you have a better idea?

**Comment by bondjimbond** [comment]
For some reason, pulling the default DSIDs into a function results in some sort of change to the characters - it won't explode on \n anymore.

**Comment by bondjimbond** [comment]
TIL that "\n" is only recognized if you double quote it.

**Comment by mjordan** [comment]
@bryjbrown the Github commenting UI if failing me.... I can't figure out how to respond inline to your [comment](https://github.com/mjordan/islandora_usage_stats_callbacks/pull/1#discussion_r141591330) about using `module_load_include()` but what you say sounds good to me :+1:  - I recommended something similar in the previous comment.

File: islandora_usage_stats_callbacks.module (hunk 2 above)
Comment (bondjimbond) [reply islandora_usage_stats_callbacks.module]: Is done!

File: islandora_usage_stats_callbacks.module (hunk 1 above)
Comment (bondjimbond) [reply islandora_usage_stats_callbacks.module]: Apparently the list() function works drastically differently between php5.5 and php7 - items are apparently keyed in opposite order. So using list() could have unpredictable results here. Stuck with foreach() but gave it a better variable name.

**Comment by mjordan** [comment]
Sorry @bondjimbond and @bryjbrown, I just now began my review, so some of my comments were "pending".

File: islandora_usage_stats_callbacks.module [hunk 3]
```diff
@@ -102,24 +102,19 @@ function islandora_usage_stats_callbacks_get_object_ids($pid) {
...
- 
+
+  module_load_include('inc', 'islandora_usage_stats_callbacks', 'includes/utilities');
+  $default_dsids = islandora_usage_stats_callbacks_set_default_dsids();
+  $dsids = variable_get(islandora_usage_stats_callbacks_cmodel_dsids, $default_dsids);
+  $dsid = array();
+  $cmodel_dsids = array();
+
+  // Convert configured DSIDs into an array
+  foreach (explode("\n", $dsids) as $pair) {
```
[the remark this answers is not in the document]
Comment (bondjimbond) [reply islandora_usage_stats_callbacks.module]: Seems to work! Commit pushed.

**Comment by bryjbrown** [comment]
@mjordan merge away, friendo.

File: includes/admin.form.inc [hunk 4]
```diff
@@ -23,6 +23,43 @@ function islandora_usage_stats_callbacks_admin_form() {
...
+  $form['content_model_wrapper']['islandora_usage_stats_callbacks_cmodel_dsids'] = array(
+    '#type' => 'textarea',
+    '#title' => t('Download DSIDs to track'),
+    '#rows' => 10,
+    '#default_value' => variable_get('islandora_usage_stats_callbacks_cmodel_dsids', $default_dsids),
+    '#description' => t("Place each CModel and DSID pair, separated by a |, on its own line. Enter FALSE for CModels with no downloadable datastreams."),
+  );
+
+
   $form['save_configuration'] = array(
```
[the remark this answers is not in the document]
Comment (mjordan) [reply includes/admin.form.inc]: @bondjimbond the new form value isn't persisting. If you add the following line as the last line in `islandora_usage_stats_callbacks_admin_form_submit($form, &$form_state)`, it will:

```
variable_set('islandora_usage_stats_callbacks_cmodel_dsids', $form_state['values']['islandora_usage_stats_callbacks_cmodel_dsids']);
```
